src/out_ref.rs

```diff
@@ -0,0 +1,573 @@
+                self.assume_init(),
+            )
+        }
+    }
+
+    #[inline]
+    pub
+    fn as_mut_ptr (self: &'_ mut Out<'out, T>)
+      -> *mut T
+    {
+        self.0.as_ptr()
+    }
+
+    #[inline]
+    pub
+    unsafe
+    fn assume_init (mut self: Out<'out, T>)
+      -> &'out mut T
+    {
+        &mut *self.as_mut_ptr()
+    }
+
+    /// # Safety
+    ///
+    /// The obtained reference cannot be used to write garbage
+    /// (`MaybeUninit::uninit()`) into the pointee.
+    ///
+    /// # Counter-example
+    ///
+    /// The following code is Undefined Behavior
+    ///
+    /// ```rust,no_run
+    /// use ::uninit::prelude::*;
+    ///
+    /// let mut my_box = Box::new(42);
+    /// let at_my_box: Out<'_, Box<i32>> =
+    ///     my_box
+    ///         .manually_drop_mut()
+    ///         .as_out()
+    /// ;
+    /// // Overwrite `my_box` with uninitialized bytes / garbage content.
+    /// unsafe {
+    ///     *at_my_box.as_uninit_mut() = MaybeUninit::uninit();
+    /// }
+    /// // Runs the destructor for a `Box<i32>` using a garbage pointer that
+    /// // may thus point anywhere in memory!
+    /// drop(my_box)
+    /// ```
+    #[inline]
+    pub
+    unsafe
+    fn as_uninit_mut (self: Out<'out, T>)
+      -> &'out mut MaybeUninit<T>
+    {
+        &mut *({self}.as_mut_ptr().cast())
+    }
+}
+
+impl<'out, T : 'out> Default for Out<'out, [T]> {
+    #[inline]
+    fn default ()
+      -> Self
+    {
+        <&mut [MaybeUninit<T>]>::into(&mut [])
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [T]> for Out<'out, [T]>
+where
+    T : Copy,
+{
+    #[inline]
+    fn from (slice: &'out mut [T])
+      -> Out<'out, [T]>
+    {
+        Out(
+            slice.into(),
+            Default::default(),
+        )
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [ManuallyDrop<T>]> for Out<'out, [T]> {
+    #[inline]
+    fn from (slice: &'out mut [ManuallyDrop<T>])
+      -> Out<'out, [T]>
+    {
+        unsafe {
+            // # Safety
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            Out(
+                ptr::NonNull::new_unchecked(
+                    slice as *mut [ManuallyDrop<T>] as *mut [T]
+                ),
+                Default::default(),
+            )
+        }
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [MaybeUninit<T>]> for Out<'out, [T]> {
+    #[inline]
+    fn from (slice: &'out mut [MaybeUninit<T>])
+      -> Out<'out, [T]>
+    {
+        unsafe {
+            Out(
+                ptr::NonNull::new_unchecked(
+                    slice as *mut [MaybeUninit<T>] as *mut [T]
+                ),
+                Default::default(),
+            )
+        }
+    }
+}
+
+impl<'out, T : 'out> Out<'out, [T]> {
+    #[inline]
+    pub
+    fn as_mut_ptr (self: &'_ mut Self)
+      -> *mut T
+    {
+        self.0.as_ptr() as *mut T
+    }
```

Comment:
`pub fn as_ptr() -> * const T` seems useful for pointer const-correctness, which may avoid the odd bug.